Run-time log filtering via vtysh commands

sworleys opened the pull request:
These patches provide a way to filter logs on arbitrary strings. If you add a filter and the log does not match it, it will be dropped. This should assist with live debugging of specific prefixes or labels, etc. when you don't want to turn on a bunch of debugs and it slow FRR to a halt.

1) debug zebra rib
2) log-filter *prefix*
3) sharp install 300,000 routes

Commands have been cleaned up since this video but it is still roughly the same:
https://asciinema.org/a/9mg6IfxgnNm5cPd7CuMoGqnHE

sworleys commented:
Force pushed to fix SA warnings

qlyoung commented on lib/log.c:
```diff
@@ -65,6 +65,103 @@ const char *zlog_priority[] = {
+	pthread_mutex_lock(&loglock);
+
+	if (zlog_filter_count >= ZLOG_FILTERS_MAX) {
+		pthread_mutex_unlock(&loglock);
```
I would suggest having `int ret` that you set and then a goto to the unlock at the end of the function instead of having an unlock along multiple return paths

qlyoung commented on lib/log.c:
```diff
@@ -65,6 +65,103 @@ const char *zlog_priority[] = {
+	strlcpy(zlog_filters[zlog_filter_count], filter,
+		sizeof(zlog_filters[0]));
+
+	if (zlog_filters[zlog_filter_count] == NULL
```
This will never be NULL, it's static memory

qlyoung commented on lib/log.c:
```diff
@@ -65,6 +65,103 @@ const char *zlog_priority[] = {
+
+	if (found_idx == -1) {
+		/* Didn't find the filter to delete */
+		pthread_mutex_unlock(&loglock);
```
Could use same goto pattern here if so inclined

qlyoung commented on lib/log.c:
```diff
@@ -65,6 +65,103 @@ const char *zlog_priority[] = {
+	/* Remove and adjust the filter array */
+	for (int i = found_idx; i < zlog_filter_count - 1; i++)
+		strlcpy(zlog_filters[i], zlog_filters[i + 1],
+			sizeof(zlog_filters[0]));
```
You can do this with a single memmove

sworleys replied:
Ah, memmove can be executed on overlapping strings, nice.

qlyoung commented on lib/log.c:
```diff
@@ -65,6 +65,103 @@ const char *zlog_priority[] = {
+			       zlog_filters[i]);
+		len += ret;
+		if ((ret < 0) || ((size_t)len >= max_size)) {
+			pthread_mutex_unlock(&loglock);
```
Use a goto

qlyoung commented on lib/log.c:
```diff
@@ -202,42 +314,104 @@ static void vzlog_file(struct zlog *zl, struct timestamp_control *tsctl,
+	return -1;
+}
+
+/* Filter out a log */
```
Probably add a note that says this must be called with loglock held

qlyoung commented on lib/log.c:
```diff
@@ -65,6 +65,103 @@ const char *zlog_priority[] = {
+static char zlog_filters[ZLOG_FILTERS_MAX][ZLOG_FILTER_LENGTH_MAX + 1];
+static uint8_t zlog_filter_count;
+
+static int zlog_filter_lookup(const char *lookup)
```
Add a node that says must be called w/ loglock held

qlyoung commented on lib/log.c:
```diff
@@ -178,17 +275,32 @@ size_t quagga_timestamp(int timestamp_precision, char *buf, size_t buflen)
 
-/* Utility routine for current time printing. */
-static void time_print(FILE *fp, struct timestamp_control *ctl)
+static void timestamp_control_render(struct timestamp_control *ctl)
```
inline this

qlyoung commented on lib/log_vty.c:
```diff
@@ -0,0 +1,106 @@
+	int ret = 0;
+
+	if (!filter) {
+		vty_out(vty, "\tno filter string given\n");
```
Delete this whole block, filter is always non-null

qlyoung commented on lib/log_vty.c:
```diff
@@ -0,0 +1,106 @@
+		ret = zlog_filter_add(filter);
+
+	if (ret == 1) {
+		vty_out(vty, "\tfilter table full\n");
```
Not \t, use `%%` followed by a space.

qlyoung commented on lib/log_vty.c:
```diff
@@ -0,0 +1,106 @@
+		vty_out(vty, "\tfilter table full\n");
+		return CMD_WARNING;
+	} else if (ret != 0) {
+		vty_out(vty, "\tfailed to %s log filter\n",
```
Same here

qlyoung commented on lib/log_vty.c:
```diff
@@ -0,0 +1,106 @@
+		return CMD_WARNING;
+	}
+
+	vty_out(vty, "\t%s\n", filter);
```
One space, not a tab pls

qlyoung commented on lib/log_vty.c:
```diff
@@ -0,0 +1,106 @@
+       FILTER_LOG_STR)
+{
+	zlog_filter_clear();
+	vty_out(vty, "\tcleared all filters\n");
```
No output necessary here

qlyoung commented on lib/log_vty.c:
```diff
@@ -0,0 +1,106 @@
+	len = zlog_filter_dump(log_filters, sizeof(log_filters));
+
+	if (len == -1) {
+		vty_out(vty, "\tfailed to get filters\n");
```
`%%`

qlyoung commented on lib/log_vty.h:
```diff
@@ -0,0 +1,26 @@
+#ifndef __LOG_VTY_H__
+#define __LOG_VTY_H__
+extern void log_filter_cmd_init(void);
+#endif
```
`#endif /* __LOG_VTY_H__ */`

qlyoung commented on lib/log_vty.h:
```diff
@@ -0,0 +1,26 @@
+ * Copyright (C) 2019 Cumulus Networks, Inc.
+ *                    Stephen Worley
+ *
+ * This file is part of GNU Zebra.
```
No, it's not :)

Use the copyright header in the developer's docs

eqvinox commented on lib/log.c:
```diff
@@ -202,42 +314,104 @@ static void vzlog_file(struct zlog *zl, struct timestamp_control *tsctl,
+		return search_buf(buf);
+
+	va_copy(ac, args);
+	ret = vsnprintf(buf + len, sizeof(buf) - len, format, ac);
```
`vsnprintfrr` - but superseded by next comment.

eqvinox commented on lib/log.c:
```diff
@@ -202,42 +314,104 @@ static void vzlog_file(struct zlog *zl, struct timestamp_control *tsctl,
+
+	/* If it doesn't match on a filter, do nothing with the log */
+	if (zlog_filter_count
+	    && vzlog_filter(zl, &tsctl, proto_str, priority, format, args)) {
```
NAK on passing `format, args` here - that means we do the formatting twice, incurring twice the cost.  Just move up the `vasnprintfrr()` below if anything.

donaldsharp commented:
So I agree that a lot of debug messages give insufficient information to know `whom` they belong to.  But on the other hand if we have complex interactions going on between multiple daemons this will be a god-send in our ability to cut the log file down to a more manageable size.  I've been thinking for a long time about requiring all log messages to have context clues about what they belong to.  It might be a good time to start scouring the messages to make them more useful for both grep'ing and reducing log size to what is important.

sworleys commented:
@eqvinox
> What exactly are you trying to do here? The `snprintf` cost is still incurred, only the `write()` or `syslog()` is saved. Are you trying to reduce disk load and usage?
>

So basically, I am trying to provide a way that people can turn on a bunch of debugs and actually be able to use them. With even a handul of debugs on, FRR will grind to a halt in scaled environments. We have seen this internally with some of our customers. I have stuff I am working on long term to address this with a ring buffer, but in the mean time this allows people to filter for what they are looking for whilst having a bunch of debugs on and not have to worry about excessive logging or FRR grinding to a halt.

Indeed snprintf is still there, but the cpu cycles we save from the write() and syslog() are still very significant. If you watch the video I included you can see a pretty big difference (~60%) in route install/removal times with a filter applied vs without one applied.


These are times without a filter applied with the same `debug zebra rib` as in the video and that is not even an excessively called debug:
```
[root@alfred sworley]# /usr/lib/frr/sharpd --log stdout --log-level debug                                                                                                                                                                                                                                                     
2019/06/10 16:40:03 SHARP: sharpd 7.2-dev starting: vty@2614                                                                                                                                                                                                                                                                  
2019/06/10 16:40:24 SHARP: Inserting 300000 routes                                                                                                                                                                                                                                                                            
2019/06/10 16:41:22 SHARP: Installed All Items 58.371050                                                                                                                                                                                                                                                                      
2019/06/10 16:42:03 SHARP: Removing 300000 routes                                                                                                                                                                                                                                                                             
2019/06/10 16:43:02 SHARP: Removed all Items 58.907280                                                                                                                                                                                                                                                                        
2019/06/10 16:43:14 SHARP: Terminating on signal 
```

srimohans commented:
@sworleys 
One question:

When I switch on the filter, and after logging for some time, I assume we will loose all the logs that were not passed the filter criterion. For example, after configuring a filter say 10.1.1.1/24 and after 5 minutes, all the logs that I will have are only with string 10.1.1.1/24 present and all the other logs are gone. logged neither to syslog nor to any other. 

Is this right understanding ?

sworleys commented:
> @sworleys
> One question:
> 
> When I switch on the filter, and after logging for some time, I assume we will loose all the logs that were not passed the filter criterion. For example, after configuring a filter say 10.1.1.1/24 and after 5 minutes, all the logs that I will have are only with string 10.1.1.1/24 present and all the other logs are gone. logged neither to syslog nor to any other.
> 
> Is this right understanding ?

Yes, those logs will be lost. The active choice in this PR is to ignore those logs to save processing time and only focus on filtered ones.

I do think it is a good idea to filter only on DEBUG messages, so I will update this patch to retain the others types regardless of filter.

sworleys commented:
> comment from call: I like this addition and think it should go in, but have on concern that may require a documentation notation - if there is a performance impact to using filter+debug vs _no_ debug this should be document so that users are are of the impact. (of course filter+debug will be much better than debug alone, but this isn't what I'm concerned / asking about.)

Totally agree, I will run that today and post the results here. If I see a performance hit, I will document that and then give a warning in the vty output?

sworleys commented:
@louberger 

`No filter times:`
```
2019/06/14 13:16:15 SHARP: Inserting 300000 routes                                                                                                                                                                                                                                                                            
2019/06/14 13:16:35 SHARP: Installed All Items 20.183507                                                                                                                                                                                                                                                                      
2019/06/14 13:16:51 SHARP: Removing 300000 routes                                                                                                                                                                                                                                                                             
2019/06/14 13:17:14 SHARP: Removed all Items 22.616951 
```

`With filter times:`
```
2019/06/14 13:17:58 SHARP: Installed All Items 19.980025                                                                                                                                                                                                                                                                      
2019/06/14 13:18:04 SHARP: Removing 300000 routes                                                                                                                                                                                                                                                                             
2019/06/14 13:18:28 SHARP: Removed all Items 24.579 
```


These results make sense because we don't do much logging under normal operating conditions anyway. Is there a specific scenario you would like me to test though?

qlyoung commented on lib/log.c:
```diff
@@ -202,42 +314,104 @@ static void vzlog_file(struct zlog *zl, struct timestamp_control *tsctl,
+	va_end(ac);
+
+	len += ret;
+	if ((ret < 0) || ((size_t)len >= sizeof(buf)))
```
This seems like a pointless check since we `return search_buf(buf)` regardless right?

sworleys replied:
yep

louberger commented:
Are the 'with filter' times collected with 'debug logging on' (i.e.,
with the filters actually doing anything? or not?  if run without
verbose debugging+filter, then please rerun with both configured.

Thanks!

On 6/18/19 1:38 PM, Stephen Worley wrote:
> @louberger <https://github.com/louberger>
> 
> |No filter times:|
> 
> |2019/06/14 13:16:15 SHARP: Inserting 300000 routes 2019/06/14 13:16:35
> SHARP: Installed All Items 20.183507 2019/06/14 13:16:51 SHARP: Removing
> 300000 routes 2019/06/14 13:17:14 SHARP: Removed all Items 22.616951 |
> 
> |With filter times:|
> 
> |2019/06/14 13:17:58 SHARP: Installed All Items 19.980025 2019/06/14
> 13:18:04 SHARP: Removing 300000 routes 2019/06/14 13:18:28 SHARP:
> Removed all Items 24.579 |
> 
> These results make sense because we don't do much logging under normal
> operating conditions anyway. Is there a specific scenario you would like
> me to test though?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/FRRouting/frr/pull/4524?email_source=notifications&email_token=ABSFNSFMBAM7RJ7ANAUPV53P3EMQ3A5CNFSM4HYANXCKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODX7MW7I#issuecomment-503237501>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABSFNSHRS7SK6R5CF3YMY73P3EMQ3ANCNFSM4HYANXCA>.
>

sworleys commented:
I went ahead re-did all of them and just put all the results in one message to be more clear.
@louberger does this include everything you want?

This is all with logs to `stdout` and `debug zebra rib` where `debug=yes`


`debug=no, filter=no (normal operation):`
```
2019/06/14 15:46:42 SHARP: Inserting 300000 routes                                                                                                                                                                                                                                                                            
2019/06/14 15:47:05 SHARP: Installed All Items 23.46728                                                                                                                                                                                                                                                                       
2019/06/14 15:47:36 SHARP: Removing 300000 routes                                                                                                                                                                                                                                                                             
2019/06/14 15:48:00 SHARP: Removed all Items 24.89119
```

`debug=no, filter=yes:`
```
2019/06/14 15:49:42 SHARP: Inserting 300000 routes                                                                                                                                                                                                                                                                            
2019/06/14 15:50:04 SHARP: Installed All Items 21.408987                                                                                                                                                                                                                                                                      
2019/06/14 15:50:39 SHARP: Removing 300000 routes                                                                                                                                                                                                                                                                             
2019/06/14 15:51:03 SHARP: Removed all Items 24.68816 
```

`debug=yes, filter=no:`
```
2019/06/14 15:54:57 SHARP: Inserting 300000 routes                                                                                                                                                                                                                                                                            
2019/06/14 15:55:48 SHARP: Installed All Items 51.316696                                                                                                                                                                                                                                                                      
2019/06/14 15:56:43 SHARP: Removing 300000 routes                                                                                                                                                                                                                                                                             
2019/06/14 15:57:35 SHARP: Removed all Items 52.852198  
```

`debug=yes, filter=yes:`
```
2019/06/14 15:58:55 SHARP: Inserting 300000 routes                                                                                                                                                                                                                                                                            
2019/06/14 15:59:18 SHARP: Installed All Items 23.295870                                                                                                                                                                                                                                                                      
2019/06/14 15:59:32 SHARP: Removing 300000 routes                                                                                                                                                                                                                                                                             
2019/06/14 16:00:00 SHARP: Removed all Items 27.563041                                                                                                                                                                                                                                                                        
```

sworleys commented:
Force pushed to collapse the commits a bit.

sworleys commented:
rebased and forced push

sworleys commented:
@eqvinox @louberger any more input here?

louberger commented:
it looks like filtering+debug slows things down by ~10% - I think it worth mentioning this in the user docs. (could also mention that without filters, things slow down by ~100%)

sworleys commented:
> it looks like filtering+debug slows things down by ~10% - I think it worth mentioning this in the user docs. (could also mention that without filters, things slow down by ~100%)

Added some docs for this. I didn't want to use any numbers because that's dependent on the debugs you are using.

sworleys commented:
@eqvinox Any update here?

eqvinox commented:
Sorry about the delay :/
